### ISSUE TITLE
Idiomatic F# API

### ISSUE THREAD

**Issue by mikhailshilkov** [pulumi/pulumi#3644]
Pulumi .NET SDK supports F# as one of the languages to author infrastructure programs. It works quite fine, but there are several complications which make F# programs harder to write.

A typical Pulumi program consists of a number of value assignments, each calling a constructor of a resource type with proper arguments. The resource is defined by virtue of calling its constructor: the value doesn't need to be returned or passed to anywhere.

Let's look at two existing examples:

**A. Azure resources ([full code](https://github.com/pulumi/examples/blob/master/azure-fs-appservice/Program.fs)):**

``` fsharp
let resourceGroup = ResourceGroup "appservice-rg"

let storageAccount =
    Account("sa",
        AccountArgs
           (ResourceGroupName = io resourceGroup.Name,
            AccountReplicationType = input "LRS",
            AccountTier = input "Standard"))

let sku = PlanSkuArgs(Tier = input "Basic", Size = input "B1")
let appServicePlan = 
    Plan("asp", 
        PlanArgs
           (ResourceGroupName = io resourceGroup.Name,
            Kind = input "App",
            Sku = input sku))
```

**B. Kubernetes deployment ([full code](https://github.com/pulumi/templates/blob/master/kubernetes-fsharp/Program.fs)):**

``` fsharp
let deployment = 
    Pulumi.Kubernetes.Apps.V1.Deployment("nginx",
      DeploymentArgs
        (Spec = input (DeploymentSpecArgs
          (Selector = input (LabelSelectorArgs(MatchLabels = appLabels)),
           Replicas = input 1,
           Template = input (
             PodTemplateSpecArgs
              (Metadata = input (ObjectMetaArgs(Labels = appLabels)),
               Spec = input (
                  PodSpecArgs
                    (Containers = 
                      inputList [
                        input (
                          ContainerArgs
                            (Name = input "nginx",
                             Image = input "nginx",
                             Ports = 
                              inputList [
                                input (
                                 ContainerPortArgs
                                   (ContainerPortValue = input 80))]))]))))))))
```

Here are some challenges:

### 1. Inputs and outputs.

Resources may contain output values: e.g. `resourceGroup.Name` in the first snippet. Its value might not be known at the time when this property is used, so a special type `Output<T>` is used to define the outputs. It's somewhat similar to `Task<T>`.

Each resource's constructor accepts a property bag where each property typically has type `Input<T>` or a type derived from it. An input can be a plain value (`1`, `"nginx") or point to an output value of another resource. This allows chaining the resources and creating dependencies between them.

While C# has implicit conversions between plain values, outputs, and inputs, F# type system is strict and requires conversions. Helper functions `input`, `io`, `inputList`, etc. were created to simplify the conversions, but they still clutter the code, and might be tricky to get right at times.

### 2. Indentation.

F# is white space significant, so the indentations must be right. The kubernetes example above shows that some resource args may use subtypes and sub-sub-types etc, so indentation gets quite heavy. It's also very strict, and compiler messages are not very helpful in getting it right.

Going one level deeper requires a `(` or `[`, so the final statement ends with something like `))]))]))))))))`.

---

For inspiration, a comparable kubernetes example in TypeScript looks much more concise:

``` ts
const deployment = new k8s.apps.v1.Deployment("nginx", {
    spec: {
        selector: { matchLabels: appLabels },
        replicas: 1,
        template: {
            metadata: { labels: appLabels },
            spec: { containers: [{ name: "nginx", image: "nginx" }] }
        }
    }
});
```

The goal is to bring readability and write-ability of F# programs close to this level.

---

Good places to give a try to a Pulumi program in F#:

- [Getting Started with Kubernetes](https://www.pulumi.com/docs/get-started/kubernetes/install-pulumi/) - use `kubernetes-fsharp` template when asked ([link to the template](https://github.com/pulumi/templates/tree/master/kubernetes-fsharp)]
- [Azure AppService example](https://github.com/pulumi/examples/tree/master/azure-fs-appservice)

**Comment by mikhailshilkov** [comment]
One possible avenue could be relying on computation expressions with custom operations. The kubernetes deployment example above could look like:

``` fsharp
  let deployment = deployment "nginx" {
    spec (
      deploymentSpec {
        replicas 1
        selector (LabelSelectorArgs(MatchLabels = appLabels))
        template (
          podTemplateSpec {
            metadata (ObjectMetaArgs(Labels = appLabels))
            spec (
              podSpec {
                containers [
                  container {
                    name "nginx"
                    image "nginx"
                    ports [ContainerPortArgs(ContainerPortValue = input 80)]}]})})})
  }
```

While it's more lightweight, it still has many levels and a `)]}]})})})` at the end. Besides, this solution requires generation of custom F#-specific assemblies for each Pulumi provider, which is possible but not very compelling.

**Comment by vshapenko** [comment]
Take a look on similar problem in Fabulous.  Here is a little sample: 
 `    let draw (view: ViewElement) navBar=
          View.Grid(
            rowdefs = [Dimension.Star; Dimension.Auto],
            coldefs = [Dimension.Star],
            rowSpacing = 0.,
            columnSpacing = 0.,
            children = [
              yield view |> row 0
              match navBar with
              | Some bar -> yield bar |> row 1
              | None -> ()
            ]
          )`

**Comment by Lanayx** [comment]
CE makes it much more attractive, the braces can be easily converted to js-way. Still one thing that doesn't look ok is repeating _spec (_ elements, without it things should look better. Is it possible to shorten it to the template below?
```
  let deployment = deployment "nginx" {
      deploymentSpec {
        replicas 1
        selector (LabelSelectorArgs(MatchLabels = appLabels))
        template (
              metadata (ObjectMetaArgs(Labels = appLabels))
              podSpec {
                containers [
                  container {
                    name "nginx"
                    image "nginx"
                    ports [ContainerPortArgs(ContainerPortValue = input 80)]
                  }
                ]
              }
        )
      }
  }
```

**Comment by mikhailshilkov** [comment]
@Lanayx Agreed! `template` is a custom operation of `deploymentSpec`, while `podTemplateSpec` is another computation expression. I haven't found a way to merge them - do you know how?

**Comment by mikhailshilkov** [comment]
@vshapenko In your example, there are no nested type instantiations. In Pulumi, every resource constructor (`Something`) accepts and argument bag (`SomethingArgs`), which may, in turn, contain sub-args (`SomethingElseArgs`). That's how cloud resources are modeled, but this leads to nesting problems.

I don't think Fabulous has anything akin to inputs-outputs either.

**Comment by Lanayx** [comment]
@mikhailshilkov looks like it's possible, have you taken a look into http://www.fssnip.net/hf/title/DSL-for-constructing-HTML ?

**Comment by Szer** [comment]
Also you could take a look at how Saturn deals with nested CE here

https://github.com/SaturnFramework/Saturn/blob/master/sample/Controller.Sample/Controller.Sample.fs

**Comment by mikhailshilkov** [comment]
@Lanayx Thanks for the link, it's a good one.

Although I was able to construct a kubernetes deployment with the syntax from your first message (i.e. similar to HTML, no spec-spec redundant pairs), I'm not yet able to make it useful.

HTML is different in an important way: any container can be a parent of any other element. So, the code keeps track of the current path with a stack.

Pulumi components aren't homogenous: I want to enforce the exact layering structure, e.g. deployment spec -> template -> pod spec -> container. There should be exactly one element at the exact place in the hierarchy.

I can't find a way to enforce this yet. How should I link a pod spec to a template with a 1-on-1 linked relationship?

**Comment by mikhailshilkov** [comment]
@Szer Do you mean this snippet?

```fs
controller {
    index (fun _ -> task {
        return {a = "hello"; b = "world"}
    })
}
```

It seems to be similar to my [comment above](https://github.com/pulumi/pulumi/issues/3644#issuecomment-564560677) and seems to have the same problems of rather complicated nesting (the use of `index` function, not a CE directly inside another CE).

**Comment by Lanayx** [comment]
Yes, I can't the way to enforce restriction either. Found an closed proposal https://github.com/fsharp/fslang-suggestions/blob/d48c35ce216e2bff148937ec028ad61e5c273fdf/archive/suggestion-7199500-allow-nested-computation-expression-for-writing-b.md , probably current problem will be a good killer example for @dsyme :)

**Comment by Szer** [comment]
> @Szer Do you mean this snippet?
> 
> ```fsharp
> controller {
>     index (fun _ -> task {
>         return {a = "hello"; b = "world"}
>     })
> }
> ```
> 
> It seems to be similar to my [comment above](https://github.com/pulumi/pulumi/issues/3644#issuecomment-564560677) and seems to have the same problems of rather complicated nesting (the use of `index` function, not a CE directly inside another CE).

Here is an example of type checking with Nested CE
https://gist.github.com/Szer/be85a58ef239820dd80d75e1e119492f

**Comment by mikhailshilkov** [comment]
Thanks @Szer! Based on your `Yield`-based examples, I managed to make the following snippet compile:

```fs
let d = deployment "mydep" {
    deploymentSpec {
        replicas 1
        selector {
            matchLabels appLabels
        }
        podTemplateSpec {
            metadata {
                labels appLabels
            }
            podSpec {
                container {
                    name "nginx"
                    image "nginx"
                    containerPort { 
                        containerPortValue 80
                    }
                }
            }
        }
    }
}
```

**Comment by absolutejam** [comment]
I was working on a similar CE-based builder pattern for raw Kubernetes types (So I could seriaise it to yaml/json) until I found Pulumi. For Pulumi, I tried the .net (C#) library and ultimately went with TypeScript as it was much more readable & forgiving, yet still type-safe.

In regards to the CE pattern, I do think this is the most user friendly, and there's already some work here https://github.com/UnoSD/Pulumi.FSharp.Extensions (Although, I tried to use the Kubernetes package and had no joy). 

However, I do find that this approach has the trade-off in that it obscures the type signatures of the type being built as each field has to essentially be optional as it's added in each pass (yield).

Additionally, this approach poses the risk of name clashes as this would have to use functions (not custom operations), with things like metadata name and container name. Although, this could be alleviated by making aliases or fully-qualifying them (eg. `Metadata.name` and `Container.name`).

If the CE approach was the way forward, I'd suggest each function should be in the form `'t -> 't` so that it's easy to extend and make abstractions.

```fsharp
// Assuming `Container` is a record and the CE folds over each function on build
let name value (container: Container) = { container with name = value }
let image value (container: Container) = { container with image = value }

container {
    name "nginx"
    image "image"
}
```

**Comment by UnoSD** [comment]
@absolutejam for K8s support, have a look at Tomas' fork: https://github.com/UnoSD/Pulumi.FSharp.Extensions/pull/3 I have not had a chance to merge it, but it should work if you get the source and compile it yourself.
